### ISSUE TITLE
createFromImmutable is not returning Carbon instance as phpDoc says

### ISSUE THREAD

**Issue by dmitryuk** [briannesbitt/Carbon#2104]
Hello,

I encountered an issue with the following code:
```php
return Carbon::createFromImmutable(new \DateTimeImmutable())
```

Carbon version: 2.35.0

PHP version: 7.4

I expected to get:
`Carbon` instance

But I actually get:
`Datetime` instance
Issue can be done two ways:
* fix phpdoc comment (because Carbon is not returned here)
`@method        static Carbon  createFromImmutable(DateTimeImmutable $dateTime)`
* Override method `createFromImmutable` in Carbon and return Carbon instance PR https://github.com/briannesbitt/Carbon/pull/2105.

Thanks!

**Comment by kylekatarnls** [comment]
Hello, thanks for the report.

Indeed this PHPDoc is auto-generated and `createFromImmutable` is actually the native PHP `DateTime`: https://www.php.net/manual/en/datetimeimmutable.createfromformat.php

One may expect it respects the inheritance. But sadly it doesn't.

I like your PR change, but:
1. it can be done only if the reciprocal is consistent:
```php
CarbonImmutable::createFromMutable(new DateTime())
```
2. It its a breaking-change so I also wonder if it might have a big impact. Mostly for libraries that expect a DateTime/DateTimeImmutable (ignoring it actually can be a Carbon instance) which currently run `createFromImmutable` get a `DateTime` so this change will make some small differences:
    - It will no longer throw an error if the passed date is not immutable.
    - Having `Carbon` instance instead of `DateTime` can make some small incompatibilities: the output when passed to `json_encode` is different for instance.
3. As the code used in the fix reveals it: the `instance` would do the job well. So we could just change the PHPDoc and promote in the doc `Carbon::instance()` and `CarbonImmutable::instance()` as the way to create from `DateTime` and `DateTimeImmutable`

**Comment by dmitryuk** [comment]
Thanks for your time.
1. Updated PR
2. I have faced with an issue by chance with a code below:
`\Carbon\Carbon::createFromImmutable(new \DateTimeImmutable())->subDay()`
So, IDE makes a hint to me that result of createFromImmutable is Carbon, but obviously it is not a truth.
3. To fix the situation with Mutable|Immutable and matched methods createFromImmutable|createFromMutable I created 2 traits, so please take a look to updated PR.

**Comment by kylekatarnls** [comment]
2. It's still a problem. While aligning the hint to the current behavior would have be without consequences, overriding the native PHP method to match the hint is the BC described above, this means your PR could only be merged in the `3.x` branch to be released in 3.0.0.
3. Mutable|Immutable traits does not fix my point (3). Since `createFromImmutable` will be an alias of `instance` (which is also equivalent to `new Carbon(new DateTimeImmutable())`) it actually brings no value. To keep backward-compatibility `Carbon` already has many aliases and often provide multiple ways to do the same thing. But I'm not willing to keep this way, it just makes the library heavy for minor convenience. Then if you use in Laravel `Date::createFromImmutable()` then someday you choose to switch from `Carbon` to `CarbonImmutable` (which can be a good move in many situations by the way), then `createFromImmutable` will fail as it does not exist in `CarbonImmutable`. An other point of interest of the native PHP method is it throw an exception if you pass a non-immutable date. And that's a good thing because you explicitly ask for an immutable, but if we replace it with `instance()` call, we lost this benefit of the native method.
4. Of course, we should not break declaration compatibility (this should be checked carefully with PHP 7.1 which does not handle covariance and contravariance) and it should not break potential sub-classes that yet override it. To be short: it should be an exact match.

All those consideration make me think that changing the hint + showing example of simple casts using `new Carbon(new DateTimeImmutable())` could be at least as good as overriding the methods to output Carbon instances.

**Comment by kylekatarnls** [comment]
Documentation update to mention constructor possibilities.

And PHPDoc typehint has been removed:
https://github.com/briannesbitt/Carbon/commit/d686622b34d6e69701ae39aa3d38355a2c2bf215#diff-c060a4f07dfb6bce7015fa79a21b6963L506
